### PR TITLE
Rewrite/redirect '#<topic>' search expressions to 'topic:<topic>'

### PR DESCRIPTION
**Pull request by isoos**
- This was mentioned while we have discussed #8189.
- We don't actively keep track of the topics, if we want to restrict the rewrite to existing topics, we should probably cache the topic names locally too.

### CODE REVIEW

**Review by sigurdm — APPROVED** [review]
I don't really worry that the rewrite happens for non-existing topics. That is what I would expect.

My only question is, if the rewrite should go the other way (such that `#topic` is the canonical version) @szakarias any opinion?

File: app/test/search/backend_test.dart [hunk 1]
```diff
@@ -69,5 +69,12 @@ void main() {
         'topic:widget abc',
       );
     });
+
+    test('query with topic shortcut `#`', () {
+      expect(
+        canonicalizeSearchForm(_parse('#widget #testing'))?.query,
+        'topic:widget topic:testing',
+      );
```
Comment (jonasfj) [inline app/test/search/backend_test.dart:77]: Can we have a test case where I search for `"#hash"` -- and would that work if I actually want to search for a keyword that starts with `#` ?

**Review by jonasfj — APPROVED** [review]

File: app/test/search/backend_test.dart (hunk 1 above)
Comment (isoos) [reply app/test/search/backend_test.dart:77]: `#` is by default removed from the content at indexing and from queries (used as a separator).

The end result of this change is that `#hash` search from https://pub.dev/packages?q=%23hash gets redirected to https://pub.dev/packages?q=topic%3Ahash and yeah, we can add a test for that.

**Comment by isoos** [comment]
> My only question is, if the rewrite should go the other way (such that #topic is the canonical version)

I'm 50:50 on this one. However, going to the other direction would require more change in the query parsing, and we can start with this one first, allowing the use of `#`, while still reverse the canonicalization at a later point.

**Pull request merged**